src/datasources/asset-calibration/AssetCalibrationDataSource.ts

```diff
@@ -52,17 +53,22 @@ export class AssetCalibrationDataSource extends DataSourceBase<AssetCalibrationQ
   ]);
 
   private readonly baseUrl = this.instanceSettings.url + '/niapm/v1';
+  private dependenciesLoadedPromise: Promise<void>;
```

Comment:
I think we could approach this in a way that keeps us from having to manage all this state. The caches should be implementation details that aren't exposed to "clients" (the query editor and other parts of the data source), and they should be lazy. We already have a `getWorkspaces()` method that returns cached data: https://github.com/ni/systemlink-grafana-plugins/blob/ac08a7d893e727beff1135c90a55b3502a906c2f/src/core/DataSourceBase.ts#L65

Then we can write a `async getSystemAliasMap()` method that does the same thing—makes a request the first time it's called and returns cached data after that. Then anytime we need this data, we just call those methods. In the query editor, use the `useAsync` hook for a clean way to manage the loading state: 

https://github.com/ni/systemlink-grafana-plugins/blob/8a028a82a2688d105eb2e165fcf83a558e2069ba/src/datasources/data-frame/components/DataFrameQueryEditor.tsx#L18

Reply:
I am not sure that the implemented cache is working as expected, take a look at this for example.

![image](https://github.com/user-attachments/assets/9c302fb7-b3a2-487c-b1aa-8e39397b6cfb)

We log the instance, workspaces, and a counter to track how many times the function is hit.
We take a dashboard that loads workspace-datasource 2 times and asset-calibration one time.  The setup produces the following results.

![image](https://github.com/user-attachments/assets/d703cd02-034e-440a-a0db-af50081cbff0)

We can notice that the state is shared only between same type datasource, so even if we keep this kind of cache, it will only be accessible to same type datasources. Even if this is what we want, current implementation is not really working, if the function is called too fast it will trigger multiple calls, not waiting for a stream of data that is already started.

We should take into consideration a solution that can share necessary data between different sources, maybe even invalidate and refresh the data as well. The way plugins are managed by grafana doesn't seem to allow us to do this by storing in memory data in a base class. We should talk about alternative ways to store this type of data so we can get a result like this one for example.

![image](https://github.com/user-attachments/assets/70d10d93-266b-472c-8eb1-758a77284cb8)

Taking this into consideration I don't really see a benefit in moving this logic in base, at least for now.

Reply:
@saradei-ni and I discussed this in a call. 

We agreed that the cache for `DataSourceBase.getWorkspaces()` has race conditions—if there are multiple callers before the fetch returns, we will make duplicate requests. I will fix this in a separate change and unblock this PR. In a later change we can apply the same approach for caching systems. 

There's another issue which is that plugin modules are isolated from each other, so there's not a straightforward way to share data between different plugins. So if you have 3 different data sources on a dashboard that all need workspace metadata, we would make 3 separate requests. Resolving this will take some research.